common-util/src/main/kotlin/com/google/devtools/ksp/PersistentMap.kt

```diff
@@ -0,0 +1,85 @@
+package com.google.devtools.ksp
+
+import com.intellij.util.io.DataExternalizer
+import com.intellij.util.io.IOUtil
+import com.intellij.util.io.KeyDescriptor
+import com.intellij.util.io.PersistentHashMap
+import java.io.DataInput
+import java.io.DataInputStream
+import java.io.DataOutput
+import java.io.File
+
+abstract class PersistentMap<K, V>(
+    storageFile: File,
+    keyDescriptor: KeyDescriptor<K>,
+    dataExternalizer: DataExternalizer<V>
+) : PersistentHashMap<K, V>(
+    storageFile,
+    keyDescriptor,
+    dataExternalizer
+) {
+    val keys: Collection<K>
+        get() = mutableListOf<K>().also { list ->
+            this.processKeysWithExistingMapping { key -> list.add(key) }
+        }
+
+    operator fun set(key: K, value: V) = put(key, value)
+
+    fun clear() {
+        keys.forEach {
+            remove(it)
+        }
+    }
+
+    fun flush() = force()
+}
+
+object FileKeyDescriptor : KeyDescriptor<File> {
+    override fun read(input: DataInput): File {
+        return File(IOUtil.readString(input))
+    }
+
+    override fun save(output: DataOutput, value: File) {
+        IOUtil.writeString(value.path, output)
+    }
+
+    override fun getHashCode(value: File): Int = value.hashCode()
+
+    override fun isEqual(val1: File, val2: File): Boolean = val1 == val2
+}
+
+object FileExternalizer : DataExternalizer<File> {
+    override fun read(input: DataInput): File = File(IOUtil.readString(input))
+
+    override fun save(output: DataOutput, value: File) {
+        IOUtil.writeString(value.path, output)
```

Comment:
Yes, this is guaranteed by its callers, who know the project / build roots better.

Reply:
That sounds good, but note that we will need to write **Unix-style** relative paths. If on Windows we write `a\b\c`, then the path can't be resolved back to an absolute path on Linux. I can't find the use of `invariantSeparatorsPath` in KSP, so I thought I should let you know just in case.

Just FYI, here's the current implementation in KGP:
  - [RelocatableFileToPathConverter](https://github.com/JetBrains/kotlin/blob/dd234157932c7304bedcdc64f08a04ad08cb0cc6/build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt#L22-L24)
  - [FileDescriptor](https://github.com/JetBrains/kotlin/blob/4efe7a00f11e7cd516a761618d7844d9049aa403/build-common/src/org/jetbrains/kotlin/incremental/storage/FileToPathConverter.kt#L41-L52)